SliderControl: Create new control using imported G2 Slider component

aaronrobertshaw opened the pull request:
Related:
- https://github.com/WordPress/gutenberg/issues/40507

## What?

Creates a new `SliderControl` component representing only the slider part of the `RangeControl` i.e. no number input field, no icons etc.

## Why?

This is a step toward better, more modular slider-related controls.

## How?

- Uses imported [G2 Slider component](https://github.com/ItsJonQ/g2/tree/main/packages/components/src/Slider) internally (this slider input is not exported)
- Imports `interpolate` utils from G2
- Follows current format recommended by guidelines e.g. TypeScript, Emotion, hooks/components etc.
- Leverages `BaseControl` for labelling purposes etc.
- `SliderControl` has a default `size` of 36px with an `__unstable-large` variant for the future 40px default.
- Adds customization of the thumb, track foreground and track background colors.

## Todo

- [ ] Write unit tests
- [ ] Add a component readme
- [ ] Fix mouse and hover event handling
- [ ] Fix custom tooltip positioning while at extremes of slider

## Testing Instructions

TBA...

- Fire up the Storybook and ensure the SliderControl functions appropriately
- Compare the SliderControl stories with those for the RangeControl
- Run unit tests for the Gutenberg Slider `npm run test-unit packages/components/src/slider/test`

## Screenshots or screencast <!-- if applicable -->



https://user-images.githubusercontent.com/60436221/182984757-5cb84071-5bfd-41f4-9f03-92d70b9d10ac.mp4

aaronrobertshaw committed:
First pass porting G2 Slider component

aaronrobertshaw committed:
Tweaks to add sizing and reset control spacing

aaronrobertshaw committed:
Updated snapshots after style and default value changes

aaronrobertshaw commented:
From https://github.com/WordPress/gutenberg/issues/40507

> Create a new SliderControl component. This component would basically be only the slider part of RangeControl (ie. no number input field, no icons..).
> 
> When it comes to the implementation of this component, we have a couple of choices: we could use the code from RangeControl, or instead use the approach from [the g2 Slider](https://github.com/ItsJonQ/g2/tree/main/packages/components/src/Slider). In both cases, we should then "adapt" the code to the current format recommended by the guidelines (e.g. TypeScript, Emotion, hooks/components..)

I'm currently working on the new `SliderControl` component mentioned above but figured it might not be too early to get some eyes on the import of the G2 Slider. Hopefully, it isn't too far off the mark 🤞

aaronrobertshaw committed:
Trial color customizations for slider input

ciampo commented on packages/components/src/slider/slider/hook.ts:
```diff
@@ -0,0 +1,131 @@
+	} = useContextSystem( props, 'Slider' );
+
+	const [ _value, onChange ] = useControlledValue( {
+		defaultValue: defaultValue || '50',
```
Since `50` could be out of bounds (depending on the values of `min` and `max`), should we fallback to `(max - min) / 2` instead?

aaronrobertshaw replied:
Absolutely. Done in [ee15f0576f](https://github.com/WordPress/gutenberg/pull/42315/commits/ee15f0576fe01b687828bbc9306e09cada88c1ee)

ciampo commented on packages/components/src/slider/slider/hook.ts:
```diff
@@ -0,0 +1,131 @@
+			let next = `${ nextValue }`;
+
+			if ( initialUnit ) {
+				next = createCSSUnitValue( nextValue, initialUnit );
```
Shouldn't slider only deal with unitless values?

aaronrobertshaw replied:
The G2 Slider dealt with units so I maintained that approach here. 

Only dealing with unitless values would simplify this component in favour of moving that logic to any other component that uses the `Slider` and units in its values. Is there value in keeping that logic here so it is consistent for all consumers of `Slider`?

ciampo replied:
In my views, `SliderControl` should deal with unitless values.

Any other component that components `SliderControl` together with other input controls dealing with units (e.g `UnitControl`) will have to keep the numeric part in sync with `SliderControl` internally (similarly to what `UnitControl` does internally with `NumberControl`)

ciampo commented on packages/components/src/slider/slider/hook.ts:
```diff
@@ -0,0 +1,131 @@
+		[ parseFloat( `${ min }` ), parseFloat( `${ max }` ) ],
+		[ 0, 100 ]
+	);
+	const componentStyles = { ...style, '--progress': `${ currentValue }%` };
```
Wasn't aware of this syntax — I've always used `element.style.setProperty("--property-name", propertyValue);`
Is it officially supported by all browsers?

Also, should we namespace the property name to avoid any potential conflict?

aaronrobertshaw replied:
> I've always used element.style.setProperty("--property-name", propertyValue);

I think this might just be personal preference and what's easiest in any given situation.

> Is it officially supported by all browsers?

We're just setting a CSS custom property which, in my understanding, is treated like setting any other inline style here. I've tested across Chrome, Firefox, Safari, and Edge. Looks ok to me.

ciampo replied:
Thank you for the explanation! TIL about using the object spread syntax also for CSS custom props

ciampo commented on packages/components/src/slider/slider/hook.ts:
```diff
@@ -0,0 +1,131 @@
+			error && styles.error,
+			styles[ size ],
+			isFocused && styles.focused( colors ),
+			error && isFocused && styles.focusedError,
```
Could we use CSS pseudo-selectors instead of JS logic for focused and focused error states?

aaronrobertshaw replied:
I did question this initially but ended up importing this faithfully from the G2 implementation of the Slider. 

Today, I've dug deeper into why this approach was taken. In Safari, I ran into problems getting it to reflect the focus state when solely relying on pseudo-selectors. 

It appears that there is something funky with its shadow DOM for the range input. If you toggle on the `:focus` state for the input via Safari's dev tools, you'll see the correct focus styling. Simply interacting with the slider, however, doesn't seem to result in the input matching the focused state.

This JS logic does achieve the consistent focus styling for me across browsers.

ciampo commented on packages/components/src/slider/slider/hook.ts:
```diff
@@ -0,0 +1,131 @@
+	} );
+	const [ value, initialUnit ] = parseCSSUnitValue( `${ _value }` );
+
+	const id = useFormGroupContextId( idProp );
```
Should this used across the package for all control related components? If so, can it be promoted outside of the `ui` folder?

aaronrobertshaw replied:
The `FormGroup` readme say it's experimental at the moment. I'm not sure if it is ready for wider use but wanted to stay as close to the G2 version as possible using what was already available. Would this be better addressed in a follow-up?

ciampo replied:
I gave [`FormControl`](https://github.com/WordPress/gutenberg/blob/e6e0ea8906c0fb01ea132e3386d5e42172614ba9/packages/components/src/ui/form-group/README.md) a deeper look — it seems like it partially overlaps with [`BaseControl`](https://github.com/WordPress/gutenberg/blob/c3a49add59883be4aadf5fda0014d97ac5ba79ee/packages/components/src/base-control/README.md).

I would probably switch to `BaseControl` for the time being, and later have a look at `FormGroup` and at the chance of consolidating `BaseControl` and `FormGroup` into one component (this may be also interesting / useful in the context of the recent challenge of labelling compound components that we're facing with `BorderBoxControl`)

ciampo commented on packages/components/src/utils/flow.js:
```diff
@@ -0,0 +1,55 @@
+ * @param {(FlowValue | FlowValue[])[]} args CSS values to combine.
+ * @return {string} The combined CSS string value.
+ */
+export function flow( ...args ) {
```
We should set the types via TypeScript (instead of TSdocs)

aaronrobertshaw replied:
After recent changes to this PR reducing the use of the `flow` util, I don't think there is much benefit in this over the suggested simple string interpolation. I've removed this util in https://github.com/WordPress/gutenberg/pull/42315/commits/daf1515529a287afa74dbf833492fe4b81ebc3d2

ciampo commented on packages/components/src/utils/style-mixins.js:
```diff
@@ -3,5 +3,6 @@ export { reduceMotion } from './reduce-motion';
 export { rtl } from './rtl';
 export { font } from './font';
 export { breakpoint } from './breakpoint';
+export { flow } from './flow';
```
Should we add an export for the `interpolate` function too?

aaronrobertshaw replied:
I wasn't too sure how widespread usage of the `interpolate` functions would be, so left it out to be consistent with the other math-related utils.

Instead of adding it to `style-mixins.js`, should we be exporting all math utils? If so, maybe a follow-up PR would be best?

ciampo replied:
Makes sense, this is definitely a tidy-up task that can be done in a follow-up

ciampo commented on packages/components/src/slider/styles.ts:
```diff
@@ -0,0 +1,194 @@
+ * Internal dependencies
+ */
+import { COLORS, CONFIG } from '../utils';
+import { flow } from '../utils/flow';
```
Since we're exporting `flow` directly from the style mixins, could we rewrite this line to

```suggestion
import { flow } from '../utils';
```

?

aaronrobertshaw replied:
The flow util has been removed. See https://github.com/WordPress/gutenberg/pull/42315#discussion_r926315552

ciampo commented on packages/components/src/slider/slider/hook.ts:
```diff
@@ -0,0 +1,131 @@
+import { useContextSystem, WordPressComponentProps } from '../../ui/context';
+import { useFormGroupContextId } from '../../ui/form-group';
+import { useControlledValue, useCx } from '../../utils/hooks';
+import { interpolate } from '../../utils/interpolate';
```
Same comment as the `flow` utility (once it's exported from `utils` folder):

```suggestion
import { interpolate } from '../../utils';
```

aaronrobertshaw replied:
I'd be happy to add this to the `utils` exports. Before I do, I was [wondering](https://github.com/WordPress/gutenberg/pull/42315#discussion_r926318120) if we should be exporting all the math util functions and if this should then be part of that?

ciampo replied:
Good point. Marking this as resolved in favour of https://github.com/WordPress/gutenberg/pull/42315#discussion_r925670265

ciampo commented on packages/components/src/slider/slider/hook.ts:
```diff
@@ -0,0 +1,131 @@
+		[ onFocus ]
+	);
+
+	const currentValue = interpolate(
```
It looks like a better name for this variable could be `progressPerc`
```suggestion
	// Interpolate the current value between 0 and 100, so that it can be used to
	// position the slider's thumb correctly
	const progressPerc = interpolate(
```

aaronrobertshaw replied:
I've tweaked the naming of this variable to be more fit for purpose and included the explanatory comment. See https://github.com/WordPress/gutenberg/pull/42315/commits/47ec300c1bf95fa6274cae54c211ba4f0b7acaa3.

ciampo commented on packages/components/src/slider/styles.ts:
```diff
@@ -0,0 +1,194 @@
+const getBoxShadowStyle = (
+	color: CSSProperties[ 'color' ] = COLORS.admin.theme
+) => {
+	return flow(
```
I wonder if the `flow` utility is really necessary — could we just interpolate strings instead?

aaronrobertshaw replied:
Agreed. I've removed the `flow` util in https://github.com/WordPress/gutenberg/pull/42315/commits/daf1515529a287afa74dbf833492fe4b81ebc3d2

ciampo commented on packages/components/src/slider/styles.ts:
```diff
@@ -0,0 +1,194 @@
+import { SliderColors } from './types';
+
+const getBoxShadowStyle = (
+	color: CSSProperties[ 'color' ] = COLORS.admin.theme
```
I'm wondering:
- if we should move the fallback to default value for colors in this file (thumbColor, trackColor, trackBackgroundColor) from this file to the component's hook;
- if we should assign these default values directly to the props, and therefore add the `@default` tag in the JSDocs for types

(cc @mirka )

aaronrobertshaw replied:
I think these suggestions would improve things. I've cleaned up this area in https://github.com/WordPress/gutenberg/pull/42315/commits/9acaa77cdb788d2993824b353dfc41fea366e6c9. Let me know if you have any further suggestions.

ciampo replied:
Looks good to me, but it would be still good to hear Lena's opinion on this one

mirka replied:
LGTM too 👍

ciampo commented on packages/components/src/utils/config-values.js:
```diff
@@ -27,6 +30,12 @@ const CONTROL_PROPS = {
 	controlHeightXLarge: `calc( ${ CONTROL_HEIGHT } * 1.4 )`,
 };
 
+const SLIDER_PROPS = {
```
I wonder if it makes sense to add these values (and the ones above) in a shared config, if they're going to be used only in one component.

@mirka, WDYT ?

aaronrobertshaw replied:
These were a pretty poor effort at bringing in the UI values from G2 without knowing where best they belong in Gutenberg. As part of the general color/styles clean up in https://github.com/WordPress/gutenberg/commit/9acaa77cdb788d2993824b353dfc41fea366e6c9, I've removed these.

ciampo commented on packages/components/src/utils/interpolate.ts:
```diff
@@ -0,0 +1,105 @@
+ *
+ * @return {number} The interpolated value.
+ */
+export function interpolate(
```
Should we add unit tests for `interpolate` and `interpolateRounded` ?

aaronrobertshaw replied:
I'll write some unit tests for these and have them available early next week. I've run out of time for today.

ciampo reviewed (commented):
Hey @aaronrobertshaw , thank you for working on this! I started having a preliminary look and left a few comments around.

One thing that came to mind is if this component should be wrapped with `BaseControl`, which would add support for label and help text

From what I can see, the component is already looking quite good! We may still need to implement a few features from `RangeSlider` :

- marks
- `step` prop
- labels and help text
- resetFallbackValue (although we may argue it may be a bit redundant)
- tooltip support

aaronrobertshaw commented:
Thanks for the early feedback @ciampo, it's always super helpful 🙇  

> One thing that came to mind is if this component should be wrapped with BaseControl, which would add support for label and help text
> ...
> We may still need to implement a few features from RangeSlider

My apologies for not being clearer in the PR description or my [earlier comment](https://github.com/WordPress/gutenberg/pull/42315#issuecomment-1186935266). When I noted the intent of this component to be a "low-level internal slider input", it was that this component would encapsulate just the `input[type="range"]` itself. I envisioned a `SliderControl` component adding labelling, marks, tooltips, help text etc around the `Slider` making it a real control.

Another part of my thinking was keeping this `Slider` more comparable to the G2 Slider in terms of functionality would make the review process easier and keep PRs to a more manageable size. 

A further breakdown of how I saw these slider-related controls is:

#### `Slider`
- This is the slider input itself i.e. `input[type="range"]`

#### `SliderControl`
- The main wrapper around the `Slider` input
- Provides the marks, tooltips, icons etc
- Also, accepts children which can be rendered on the same line as the input. Think `NumberControl` and reset `Button` as per `RangeControl`

#### `SliderNumberControl`
- Aims to satisfy the requirements of a `RangeControl`
- Essentially is just a convenience component that is a `SliderControl` with optional `NumberControl` and reset `Button` as children
- Has all the other internals you'd expect, e.g. triggering change handlers on either input change, resetting etc.
- This could possibly handle mapping any difference in `RangeControl` props to the new component's API to make aliasing `RangeControl` trivial

#### `SliderUnitControl`
- Similar to the `SliderNumberControl`
- Convenience component that is a `SliderControl` with optional `UnitControl` as a child.

---

I hope to have the `SliderControl` up in a PR based on this one in the next couple of days. Perhaps we can re-evaluate the general approach then unless you see any red flags now?

As for the individual feedback items in your last review, I'll work through them today.

aaronrobertshaw committed:
Ensure fallback default value is within min-max range

aaronrobertshaw committed:
Rename progress CSS custom property

aaronrobertshaw committed:
Remove flow utility in favour of simple string interpolation

aaronrobertshaw committed:
Clarify current progress percentage calculation and style

aaronrobertshaw committed:
Clean up colors, box shadows and defaults

ciampo commented:
Thank you for clarifying!

>  When I noted the intent of this component to be a "low-level internal slider input", it was that this component would encapsulate just the `input[type="range"]` itself. I envisioned a `SliderControl` component adding labelling, marks, tooltips, help text etc around the `Slider` making it a real control.
> 
> Another part of my thinking was keeping this `Slider` more comparable to the G2 Slider in terms of functionality would make the review process easier and keep PRs to a more manageable size.
> 
> A further breakdown of how I saw these slider-related controls is:
> 
> #### `Slider`
> * This is the slider input itself i.e. `input[type="range"]`
> 
> #### `SliderControl`
> * The main wrapper around the `Slider` input
> * Provides the marks, tooltips, icons etc
> * Also, accepts children which can be rendered on the same line as the input. Think `NumberControl` and reset `Button` as per `RangeControl`

After discussing with @mirka, and referring to the roadmap outlined in https://github.com/WordPress/gutenberg/issues/40507,  we don't really see the utility of having `Slider` as a separate component (i.e. living in its separate `slider` folder). At most, the `Slider` component could be a sub-component / implementation detail of the `SliderControl` component (but still not exported from the `slider-control` folder).

Also, we don't think that `SliderControl` should accept `children` to be rendered on "the same line". The `SliderControl` should basically just take care of displaying an "enhanced" slider input, implementing only the slider-related features from `RangeControl`, and not worrying about compound inputs or layouts.

The remaining features of `RangeControl`, like composing the slider input with other components (like icons,  `NumberControl`, `UnitControl`..), should be tackled by higher-level components (like the proposed `SliderNumberControl` and `SliderUnitControl`).

As we will work on these higher-level components, we will have a chance at discussing a few aspects:
- what should the internal layout options be? (e.g always display the components in a row? Or maybe on a grid? In which order? Can those be customised?)
- there are a few props / features from `RangeControl` that feel a bit "extra", like icon props and reset button. Maybe we could decide to exclude them from `SliderNumberControl` and `SliderUnitControl`, and only include them in the compat-layer that we're going to add to "translate" from `RangeControl` to `SliderNumberControl` ?
- how do we best label compound components `SliderNumberControl` and `SliderUnitControl` ? This is something that came up recently related to `BorderControl`  and `BorderBoxControl`, but the same will apply to these components too

mirka commented on packages/components/src/slider/types.ts:
```diff
@@ -0,0 +1,66 @@
+	 *
+	 * @default 'default'
+	 */
+	size?: 'small' | 'default' | 'large';
```
Ugh I'm having a hard time deciding what to do with sizes in this one. We can _kind of_ start with a clean slate. I'm thinking default=36 and __unstable-large=40. Do you think we can get away with not adding a 30px variant at all? I'm hoping so, because all our basic components now have at least a 36px variant (#39397).

Also we can basically ignore the `controlHeight*` variables in CONFIG — those are not relevant anymore and are due for removal.

aaronrobertshaw committed:
Create SliderControl relocating Slider as internal only component

Update snapshots

Remove comments

aaronrobertshaw committed:
Update use of color values

After rebasing on trunk, previous color values had been retired via #43994 and #43773

aaronrobertshaw committed:
Tweak vertical position of mark labels

aaronrobertshaw committed:
Update slider control snapshot after color changes

aaronrobertshaw committed:
Remove debounced hover interactions

This reflects the changes made in #44271 for the RangeControl.

aaronrobertshaw commented:
Just a quick update to note there is ongoing work to [clean up the NumberControl types around the value prop](https://github.com/WordPress/gutenberg/pull/45982/) that the new Slider controls would benefit from. 

I'll dust off this PR and the follow-ups for the SliderNumberControl and SliderUnitControl once #45982 lands. This most likely will not be until the new year.

ciampo commented:
Absolutely — I've been AFK for the past month. I'll finish catching up and resume work on #45982 ASAP

aaronrobertshaw commented:
Going to close this PR for now. 

It is still dependent on https://github.com/WordPress/gutenberg/pull/45982 but can be resurrected if needed.
